Add feePadding for use in balanceTransaction

Anviking opened the pull request:
- [x] Replace manually calculated padding with updating the ExUnits to max before calling `evaluateMinimumFee`.

### Comments

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-1182

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2226,14 +2235,31 @@ balanceTransaction ctx genChange (ApiT wid) body = do
                 , txTimeToLive = ttl
                 }
 
+        let realFee = evaluateMinimumFee tl pp' sealedTxIncoming
```
`realFee` is a bit confusing here, because they're still not the real fee but an estimation -- just a better one. The real fee we can obtain at the very end, by looking at the size of the serialised tx.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2214,9 +2217,15 @@ balanceTransaction
+        <$> liftIO (NW.currentNodeProtocolParameters nl)
+
+    -- TODO: Check that sealedTxIncoming contains a PParam hash matching both pp
+    -- and pp'. Without this, collateral could be forfeited!
```
I am not sure to understand this one? Note that I took care of adding the script integrity hash when necessary as part of #2952

Anviking replied:
Ah, thanks, indeed. If we always set the script integrity hash, then it should be nothing to worry about. I thought it looked like in [cardano-api](https://github.com/input-output-hk/cardano-node/blob/0ed04ab8d0a205691219f04623a3d880545a3f34/cardano-api/src/Cardano/Api/TxBody.hs#L1192) (but I now see it's only used to derive the script integrity hash), and that you said PAB would set it (https://github.com/input-output-hk/cardano-wallet/pull/2952#discussion_r723914685)

Anviking committed:
Defer more fee calculation to cardano-api

Rather than manually adding (extra) padding for
- Script integrity hash
- CBOR encoding of ExUnit

we can set ExUnits to their maximum using `updateTx` before
using the cardano-api fee estimation. This will more
automatically handle both cases.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2308,17 +2312,20 @@ balanceTransaction ctx genChange (ApiT wid) body = do
-            , transaction
-            , fee = coinToQuantity (newFee oldFee)
-            }
+    updatedTx <- liftUpdateTxErr (updateTx tl nodePParams partialTx txUpdate)
+    return $ ApiConstructTransaction
+        { coinSelection
+        , transaction = ApiT updatedTx
+        , fee = coinToQuantity (newFee oldFee)
+        }
```
Minor stylistic suggestion:

```suggestion
    let fee = coinToQuantity (newFee oldFee)
    transaction <- ApiT <$> liftUpdateTxErr (updateTx tl nodePParams partialTx txUpdate)
    return $ ApiConstructTransaction { coinSelection, transaction, fee }
```

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2352,25 +2359,29 @@ balanceTransaction ctx genChange (ApiT wid) body = do
+        -- execution units are not known yet. Calling `updateTx` like this will:
+        -- 1. make the CBOR encoding of the execution units as large as possible
+        -- 2. add the script integrity hash
+        let maxUnits = pp ^. (#txParameters . #getMaxExecutionUnits)
```
Note that the max units are already part of the context as `txPlutusScriptExecutionCost`, so it'd be better to pull from there and avoid duplicating this.

Anviking replied:
`txPlutusScriptExecutionCost` has type `Coin`, not `ExecutionUnits`, however…

Anviking replied:
I was going to say:
> I think there's some oddness to how `padFeeEstimation`, `selectAssets`, the rest of `balanceTransaction`, `TransactionCtx` all work together, but I think it's "fine"… for now / not something to be tackled by this PR.

But I suppose we could set the `newExUnits` already in `balanceTransaction` and keep the result…  not to have to do it twice… 🤔

KtorZ replied:
Actually, newExUnits is gone :s ... We'll have to discuss that.
